suitcase/utils/__init__.py

```diff
@@ -1,25 +1,230 @@
+
+    This design is inspired by Python's zipfile and tarfile libraries.
+    """
+    def __init__(self, directory):
+        self._directory = Path(directory)
+        self._reserved_names = set()
+        self._artifacts = collections.defaultdict(list)
+
+    @property
+    def artifacts(self):
+        return dict(self._artifacts)
+
+    def reserve_name(self, label, postfix):
+        """
+        Ask the wrapper for a filepath.
+
+        An external library that needs a filepath (not a handle)
+        may use this instead of the ``open`` method.
+
+        Parameters
+        ----------
+        label : string
+            partial file name (i.e. stream name)
+        postfix : string
+            relative file path and filename
+
+        Returns
+        -------
+        name : Path
+         """
+        if Path(postfix).is_absolute():
+            raise SuitcaseUtilsValueError(
+                f"The postfix {postfix!r} must be structured like a relative "
+                f"file path.")
+        name = (self._directory / Path(postfix)).expanduser().resolve()
+        if name in self._reserved_names:
+            raise SuitcaseUtilsValueError(
+                f"The postfix {postfix!r} has already been used.")
+        self._reserved_names.add(name)
+        self._artifacts[label].append(name)
+        return name
+
+    def open(self, label, postfix, mode, encoding=None, errors=None):
+        """
+        Request a file handle.
+
+        Like the built-in open function, this may be used as a context manager.
+
+        Parameters
+        -------
+        label : string
+            partial file name (i.e. stream name)
+        postfix : string
+            postfix for the filenames.
+        mode : {'x', 'xt', xb'}
+            'x' or 'xt' for text, 'xb' for binary
+        encoding : string or None
+            Passed through open.  See Python open documentation for allowed
+            values. Only applicable to text mode.
+        errors : string or None
+            Passed through to open. See Python open documentation for allowed
+            values.
+        Returns
+        -------
+        file : handle
+        """
+        filepath = self.reserve_name(label, postfix)
+        # create the directories if they don't yet exist
+        if not os.path.exists(os.path.dirname(filepath)):
+            try:
+                os.makedirs(os.path.dirname(filepath))
+            except OSError as exc:
+                if exc.errno != errno.EEXIST:
+                    raise
+
+        if mode not in ['x', 'xt', 'xb']:
+            raise ModeError(
+                f'the mode passed to MultiFileWrapper.open is {mode} but needs'
+                ' to be one of "x", "xt" or "xb"')
+        f = open(filepath, mode=mode, encoding=encoding, errors=errors)
+        return f
+
+
+class PersistentStringIO(io.StringIO):
+    ''' A version of StringIO that avoids closing the file in a context manager
```

Comment:
A StringIO that does not clear the buffer when closed or exited from a context

Reply:
fixed